### ISSUE TITLE
Number of AuthnReqeusts per sessions should be limited

### ISSUE THREAD

**Issue by baszoetekouw** [OpenConext/OpenConext-engineblock#1345]
Observed: some broken clients make >1000 AuthNRequests/minute.  This causes a huge slowdown and CPU hog, which seems to be caused by the fact that all AuthnRequests are stored in the session, and the session keeps growing to >1MB.

Suggested solution: only store the last N AuthnReqeusts in the session (which N configurable and typically in the order of 25) and discard all previous ones.

**Comment by baszoetekouw** [comment]
As discussed: of the max number of AuthnRequests is reached, return a 400 and destroy the session.

**Comment by johanib** [comment]
The 25 should be configurable.
Use status 429.
Error: too many authentication requests (on error page)

**Comment by thijskh** [comment]
@baszoetekouw You keep creating new sessions in this case then, right? So the person doing 1000 authnrequests will get 40 `429`s and still 960 `200`s each minute

**Comment by baszoetekouw** [comment]
True...  so maybe we should keep the session and only accept new AuthnRequests after a (configurable) amount of time (say, a few minutes)?  Or can you think of a more elegant solution?

The real issue btw is not the amount of requests (there is a rate limit per ip, and AuthnRequests are cheap), but rather the size of the session.

**Comment by johanib** [comment]
Discussed: Be sure to destroy the whole session when returning the 429.

**Comment by baszoetekouw** [comment]
@thijskh if the client keep retrying, they will hit the rate limit and be blocked.  The issue here was that before the client hits the rate limit, the session had grown to big already that the VM was having load issues.

**Comment by baszoetekouw** [comment]
Hmm, storing all past AuthnRequests sounds like it could become a problem anyway.  Are these sessions really never cleared server-side?  At the client side they are session-cookies, so could also be quite long lived.

I see two possible solutions then:
 - implement the `$requestStorage` as a fix-sized queue, that will automatically drop "old" AuthnRequests when new ones come in.  that woudl fix the original (rate-limiting) issue automatically.
 - implement some kind of garbage collection, that would periodically (maybe when new AuthnRequests are stored) check all stored AuthnRequests and discard "old" ones (for some definition of "old").  If then still too many are present, go with the original 429-error strategy

The second option sounds cleaner, but more involved.  The `issueInstant` parameter doesn't seem to be easily accessible in the requestStorage, and it is unclear what the lifetime of an AuthnRequest should be.
So the first options seems preferable.

**Comment by baszoetekouw** [comment]
Hmm, I think we are overlooking something.  I've looked at the session storage in production, and the oldest session it has is 5 minutes old.  So it does seem something is cleaning up old sessions?
Also, the largest session I see now has only 2 SAMLRequests stored.

**Comment by baszoetekouw** [comment]
Anyway, this means that limiting the number of stored SAMLRequests to 10-20 is probably fine.  That solved the original problem and makes sure the sessions can't grow too large (which will also help later on when we want to store the sessions in a database rather then on-disk).
